// gacp/lib/messages/history.js
"use strict";
/**
 * @since 2019-06-13 14:14
 * @author vivaxy
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.flushHistory = exports.clearHistory = exports.setHistory = exports.getHistory = exports.DEFAULT_MESSAGES = void 0;
const path = require("path");
const log = require("log-util");
const fse = require("fs-extra");
const configs_1 = require("../configs");
const { GACPHOME, HISTORY_FILE_NAME } = configs_1.PATHS;
const historyFile = path.join(GACPHOME, HISTORY_FILE_NAME);
exports.DEFAULT_MESSAGES = {
    type: '',
    scope: '',
    gitmoji: '',
    subject: '',
    body: '',
    footer: '',
};
let cache = exports.DEFAULT_MESSAGES;
function debug(...message) {
    log.debug('gacp:history', ...message);
}
function getHistory() {
    debug('reading history');
    if (cache) {
        // already read from file system
        return cache;
    }
    try {
        cache = require(historyFile);
    }
    catch (e) {
        debug('history file not exists');
        cache = exports.DEFAULT_MESSAGES;
    }
    return cache;
}
exports.getHistory = getHistory;
function setHistory(history) {
    debug(`set: ${JSON.stringify(history)}`);
    cache = { ...cache, ...history };
}
exports.setHistory = setHistory;
function clearHistory() {
    cache = exports.DEFAULT_MESSAGES;
}
exports.clearHistory = clearHistory;
async function flushHistory() {
    await fse.outputFile(historyFile, JSON.stringify(cache || {}, null, 2));
}
exports.flushHistory = flushHistory;
