// gacp/lib/configs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMOJI_TYPES = exports.ERROR_TYPES = exports.GITMOJIS = exports.PATHS = void 0;
/**
 * @since 2017-02-04 17:05
 * @author vivaxy
 */
const gitmojis_1 = require("./gitmojis");
const { HOME, HOMEPATH, USERPROFILE } = process.env;
exports.PATHS = {
    GACPHOME: `${HOME || HOMEPATH || USERPROFILE}/.gacp`,
    // from https://raw.githubusercontent.com/carloscuesta/gitmoji/master/src/data/gitmojis.json
    GITMOJI_CONFIG_FILE_NAME: 'gitmojis.json',
    // from conventional-commit-types
    COMMIT_TYPES_CONFIG_FILE_NAME: 'conventional-commit-types.json',
    HISTORY_FILE_NAME: 'history.json',
};
const { gitmojis: GITMOJIS } = gitmojis_1.default;
exports.GITMOJIS = GITMOJIS;
var ERROR_TYPES;
(function (ERROR_TYPES) {
    ERROR_TYPES["GACP"] = "gacp";
})(ERROR_TYPES = exports.ERROR_TYPES || (exports.ERROR_TYPES = {}));
var EMOJI_TYPES;
(function (EMOJI_TYPES) {
    EMOJI_TYPES["EMOJI"] = "emoji";
    EMOJI_TYPES["CODE"] = "code";
})(EMOJI_TYPES = exports.EMOJI_TYPES || (exports.EMOJI_TYPES = {}));
